### DISCUSSION TITLE
[request]: cli flag for showing compilation errors first at bottom

### DISCUSSION THREAD

**Discussion by spaceunifyfifty** [sbt/sbt#7615] [Ideas]
My typical development cycle is
1. write code
2. `sbt run` 
3. notice compilation failure. Look at error. Usually a simple fix. Go solve it.

repeat.

However, nearly 80% of the terminal output real estate is filled with low-priority information like 
```shell
[error] Total time: 0 s, completed Jul 27, 2024, 12:28:36 PM
```

For example, my terminal takes up roughly 1/3 of my screen when toggled like most IDE defaults. With that, here is all I am able to see after `sbt run`
```shell
[error]   |              ^^^^^^
[error]   |              Not found: type Foobar
[error]   |
[error]   | longer explanation available when compiling with `-explain`
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
[error] Total time: 0 s, completed Jul 28, 2024, 12:59:57 PM
```
 It is just barely cuts off before showing me the import bit: the actual problematic source code, and what file and line it's located. If I could just see that, I would know exactly where to go and what to fix.

This forces me to move my hand off of my keyboard, go to my mouse, scroll up until I see the error, then go back to homerow to fix the error. Then I re-run `sbt run`, and often have a few more compile errors to fix, forcing me to repeat this whole routine of moving my hand back and forth and scrolling. It adds up over time.

This cycle would be much less irritating to keyboard-only users if we had a flag to omit most of the first 5 low-priority lines (e.g. `--concise`) or reverse their order so the errors are at the bottom and less important things like `Total time` and `-explain` prompt are at the top (e.g. `--priority`), like so
```shell
$ sbt run --priority
[error]   | longer explanation available when compiling with `-explain`
[error] (Compile / compileIncremental) Compilation failed
[error] Total time: 0 s, completed Jul 28, 2024, 12:59:57 PM 
[error]   |
[error] -- [E006] Not Found Error: /Users/user/projects/HelloWorld/src/main/scala/Main.scala:1:14
[error] 1 |def foobar(x: Foobar) = ???
[error]   |              ^^^^^^
[error]   |              Not found: type Foobar
[error] one error found
```
or 
```shell
$ sbt run --concise
[error] (Compile / compileIncremental) Compilation failed
[error] -- [E006] Not Found Error: /Users/user/projects/HelloWorld/src/main/scala/Main.scala:1:14
[error] 1 |def foobar(x: Foobar) = ???
[error]   |              ^^^^^^
[error]   |              Not found: type Foobar
[error] one error found
```

I assume the vast majority of developers are more interested in seeing the error rather than the total time it took or optional flags for next time, so I think it would be a reasonable default to provide.

I am willing to do the PR myself if I can get some promising feedback on it's likelihood of being accepted. Thanks!

**Comment by eed3si9n** [comment]
Thanks for starting this discussion. This is an interesting request, and I think we should try to tackle it. Not to give excuse to the current UX, I think we've reached the status quo by either sbt or Scala compiler trying to inform the user that:
1. something is currently running, and/or
2. there were some problems in the code, and/or
3. how to fix the said mistakes or problem

The general trend has been to use LSP/BSP to communicate the location of the errors to the editors in a structured fashion, but I'm happy to discuss the low-tech approaches as well since they are often more reliable.

### timestamp

The timestamp can be turned off already using the following setting:

```scala
Global / showTiming := false
```

### compilation errors

Except for the "(Compile / compileIncremental) Compilation failed" line, the rest is coming from the Scala compiler itself. There's an sbt plugin written by @Duhemm called [sbt-errors-summary](https://github.com/Duhemm/sbt-errors-summary) that takes over the compiler reporter and shortens the compilation errors, including the "longer explanation available" message.

```bash
[info] Executing in batch mode. For better performance use sbt's shell
[info] compiling 1 Scala source to /private/tmp/foo/target/scala-3.3.3/classes ...
[error] [E1] src/main/scala/example/Main.scala
[error]      Not found: foobar
[error]      L6: def foo = foobar
[error]                    ^
[error] src/main/scala/example/Main.scala: L6 [E1]
[info] Legend: Ln = line n, Cn = column n, En = error n
[error] (Compile / compileIncremental) Compilation failed
```

### log level --error

Next, we can turn off all info level logs by upping the logLevel with `--error`:

```scala
$ sbt --error compile
[error] [E1] src/main/scala/example/Main.scala
[error]      Not found: foobar
[error]      L6: def foo = foobar
[error]                    ^
[error] src/main/scala/example/Main.scala: L6 [E1]
[error] (Compile / compileIncremental) Compilation failed
```

### watch command (tilde)

As a general note, using sbt as batch mode like `sbt run` slows down the compilation significantly because JVM needs to JIT the entire scala-library, scala-compiler, and all your libraries. One way to workaround it is to run the watch command `~`. This will continuously monitor the file changes and runs the passed in command, like `testQuick`. By default it will print out:

```scala
[info] 1. Monitoring source files for foo/testQuick...
[info]    Press <enter> to interrupt or '?' for more options.
```

but that can also be suppressed as:

```scala
ThisBuild / watchLogLevel := Level.Error
```

so all together we'd get:

```scala
$ sbt --error '~testQuick'
[error] [E1] src/main/scala/example/Main.scala
[error]      Not found: foobar
[error]      L6: def foo = foobar
[error]                    ^
[error] src/main/scala/example/Main.scala: L6 [E1]
[error] (Compile / compileIncremental) Compilation failed
```

which is the same as before, except, if you change the source code it will automatically run again:

```scala
$ sbt --error '~testQuick'
[error] [E1] src/main/scala/example/Main.scala
[error]      Not found: foobar
[error]      L6: def foo = foobar
[error]                    ^
[error] src/main/scala/example/Main.scala: L6 [E1]
[error] (Compile / compileIncremental) Compilation failed
[error] [E1] src/main/scala/example/Main.scala
[error]      Not found: foobar2
[error]      L6: def foo = foobar2
[error]                    ^
[error] src/main/scala/example/Main.scala: L6 [E1]
[error] (Compile / compileIncremental) Compilation failed
```
